### ISSUE TITLE
Import namespace packages?

### ISSUE THREAD

**Issue by haf** [microsoft/pylance-release#2562]
# Background:

I'm trying to use namespace packages, specifically [opentelemetry-api](https://github.com/open-telemetry/opentelemetry-python/tree/main/opentelemetry-api/src/opentelemetry) — which uses [native namespace packaging](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages)

Here's the issue as it's presented: https://github.com/open-telemetry/opentelemetry-python/issues/2591

A few years back, issue microsoft/python-language-server#288 had a lengthy discussion where some progress seems to have been made to support namespace packages, but it was closed.

Could you advice on how to best proceed? Is there a reason that you don't support namespace packages? (e.g. have they been relatively unused until now?) What are the pros and cons?

**Comment by erictraut** [comment]
Pylance does support namespace packages.

Are you seeing a specific problem? If so, please describe the symptoms, expected behavior, and repro steps.

**Comment by haf** [comment]
@erictraut https://github.com/open-telemetry/opentelemetry-python/issues/2591 -- yes, code completion doesn't work for opentelemetry and also in strict mypy mode you get errors for the package. Please click the link to see screenshots.

**Comment by erictraut** [comment]
Thanks for the additional information.

Pylance is able to resolve imports from the library, but it is not picking up the type information. There are two reasons for this. The first is a bug in the opentelemetry library. It contains inlined type information, but it is not following PEP 561, which says:

> For namespace packages (see [PEP 420](https://peps.python.org/pep-0420)), the "py.typed" file should be in the submodules of the namespace, to avoid conflicts and for clarity.

The opentelemetry library is placing the "py.typed" file in the top-level namespace directory rather than within the individual submodules. Pyright is currently looking for it in the submodule directories, consistent with PEP 561.

Even if I fix this problem by adding "py.typed" files in the individual submodules, we hit a second issue due to a bug in pyright. In strict mode, it is incorrectly generates a "stub file not found" error for the namespace portion of the import within the statement `from opentelemetry import baggage, trace`.

I've fixed the bug in pyright. I've also modified the import logic in pyright to allow for a "py.typed" file at the top level of a namespace. Even though that doesn't completely conform to the PEP 561 standard, I think it's reasonable and unambiguous. Plus, it appears that mypy accepts this variant.

This change may or may not make it into this week's release of pylance. The timing is close.

Incidentally, I ran the command-line version of pyright with the "--verifytypes" option with the opentelemetry library. Here is the output:

```
Symbols exported by "opentelemetry": 973
  With known type: 743
  With ambiguous type: 25
  With unknown type: 205

Type completeness score: 76.4%
```

That means the type information for this library is incomplete, so you may still have trouble if you try to use it with strict mode. We've prepared [this guidance](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md) to help library authors detect and address missing type information in their libraries.

**Comment by heejaechang** [comment]
This issue has been fixed in version 2022.4.1, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202241-13-April-2022)
